File: all/222_count_complete_tree_nodes/sol_better_perfect_binary_tree.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        leftmost_depth = 0
        rightmost_depth = 0

        current_node = root

        while (current_node.left):
            # Count the depth from the root node to the leftmost leaf node.
            leftmost_depth += 1

            current_node = (current_node.left)

        current_node = root

        while (current_node.right):
            # Count the depth from the root node to the rightmost leaf node.
            rightmost_depth += 1

            current_node = (current_node.right)

        if leftmost_depth != rightmost_depth:
            return (1 + (self.countNodes(root.left)) + (self.countNodes(root.right)))

        # Current binary tree is a perfect binary tree, number of nodes in the
        # current binary tree is two to the power of (tree depth plus 1) and
        # then minus - 1.
        return ((1 << (leftmost_depth + 1)) - 1)

File: all/222_count_complete_tree_nodes/sol_better_perfect_binary_tree.c
#include <stddef.h>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int countNodes(struct TreeNode* root) {
    if (NULL == root) {
        return 0;
    }

    const struct TreeNode* current_node_ptr = root;
    int leftmost_depth = 0;
    int rightmost_depth = 0;

    while (NULL != (current_node_ptr->left)) {
        /* Count the depth from the root to the leftmost leaf node. */
        ++leftmost_depth;

        current_node_ptr = (current_node_ptr->left);
    }

    current_node_ptr = root;

    while (NULL != (current_node_ptr->right)) {
        /* Count the depth from the root to the rightmost leaf node. */
        ++rightmost_depth;

        current_node_ptr = (current_node_ptr->right);
    }

    if (leftmost_depth != rightmost_depth) {
        return (1 + countNodes(root->left) + countNodes(root->right));
    }

    /*
     * Current binary tree is a perfect binary tree, number of nodes in the
     * current binary tree is two to the power of (tree depth plus 1) and then
     * minus - 1.
     */
    return ((1 << (leftmost_depth + 1)) - 1);
}
